File: packages/shared/src/ai-recognition-prompt.ts
/**
 * AI 识别提示词构造器是 Docker Go 与 Cloudflare Worker 的共同事实源。
 *
 * prompt JSON 记录可审阅的系统规则、字段规则和示例；运行时代码只注入用户上下文，
 * 避免两个后端各自维护一份不可对齐的提示词。
 */
import { z } from "zod";
import promptSpecJson from "../data/ai-recognition-prompt.json";

const promptExampleSchema = z.object({
  input: z.string(),
  output: z.unknown(),
}).strict();

const promptSpecSchema = z.object({
  version: z.string().min(1),
  schemaName: z.string().min(1),
  system: z.array(z.string().min(1)).min(1),
  outputContract: z.array(z.string().min(1)).min(1),
  fieldRules: z.array(z.string().min(1)).min(1),
  examples: z.array(promptExampleSchema).min(1),
}).strict();

export const aiRecognitionPromptSpec = promptSpecSchema.parse(promptSpecJson);
export const AI_RECOGNITION_PROMPT_VERSION = aiRecognitionPromptSpec.version;
export const AI_RECOGNITION_SCHEMA_NAME = aiRecognitionPromptSpec.schemaName;

/** 配置选项同时保留 value 与双语标签，让模型能优先复用用户已存在的分类/支付方式。 */
export interface AIRecognitionPromptConfigOption {
  value: string;
  label: string;
  zhCN: string;
  enUS: string;
}

/** 动态上下文只提供偏好与候选，不构成识别证据；模型不能凭这些选项虚构订阅。 */
export interface AIRecognitionPromptConfigContext {
  categories: readonly AIRecognitionPromptConfigOption[];
  paymentMethods: readonly AIRecognitionPromptConfigOption[];
  tags: readonly string[];
}

/** 系统 prompt 固定来自版本化 JSON，便于 Go/Worker diagnostics 回显同一个 promptVersion。 */
export function buildAIRecognitionSystemPrompt(): string {
  return aiRecognitionPromptSpec.system.join("\n");
}

/**
 * 构造单次识别 user prompt。
 *
 * 日期、时区、默认币种和现有配置都随请求注入；它们帮助模型解释输入，但不能绕过最终 schema/normalize。
 */
export function buildAIRecognitionUserPrompt({
  text,
  timezone,
  defaultCurrency,
  currentDate,
  imageCount,
  locale,
  configContext = { categories: [], paymentMethods: [], tags: [] },
}: {
  text: string;
  timezone: string;
  defaultCurrency: string;
  currentDate: string;
  imageCount: number;
  locale: string;
  configContext?: AIRecognitionPromptConfigContext;
}): string {
  const examples = aiRecognitionPromptSpec.examples.map((example, index) => [
    `Example ${index + 1} input:`,
    example.input,
    `Example ${index + 1} JSON output:`,
    JSON.stringify(example.output, null, 2),
  ].join("\n")).join("\n\n");

  const userInput = text.trim() || "(no text input; use attached images)";

  return [
    "Runtime context:",
    `- Current date in user's timezone: ${currentDate}`,
    `- User timezone: ${timezone}`,
    `- User locale: ${locale}`,
    `- Default currency hint: ${defaultCurrency}`,
    `- Attached image count: ${imageCount}`,
    "",
    "User context:",
    "Existing user tags:",
    ...formatTagsForPrompt(configContext.tags),
    "",
    "Available Renewlet configuration options:",
    "Categories:",
    ...formatConfigOptionsForPrompt(configContext.categories),
    "Payment methods:",
    ...formatConfigOptionsForPrompt(configContext.paymentMethods),
    "",
    "Task:",
    "- Extract subscriptions from the delimited user input below and any attached images.",
    "- Treat the dynamic user context as preferences and available options, not as subscription evidence.",
    "",
    "Output contract:",
    ...aiRecognitionPromptSpec.outputContract.map((rule) => `- ${rule}`),
    "",
    "Field rules:",
    ...aiRecognitionPromptSpec.fieldRules.map((rule) => `- ${rule}`),
    "",
    "Examples:",
    examples,
    "",
    "User input:",
    "<<<renewlet-user-input",
    userInput,
    ">>>",
  ].join("\n");
}

/**
 * 构造 schema repair prompt。
 *
 * Repair 只允许基于原始输入、图片、前一轮 JSON 和高置信公共知识补齐 notes，不能引入本地品牌表。
 */
export function buildAIRecognitionRepairUserPrompt({
  originalUserPrompt,
  previousObject,
  missingNoteNames,
}: {
  originalUserPrompt: string;
  previousObject: unknown;
  missingNoteNames: readonly string[];
}): string {
  return [
    "Repair task:",
    "- The previous JSON object is structurally valid but some describable subscriptions have missing or unusable notes.",
    "- Regenerate the entire JSON object with the same output contract, field rules, and all subscriptions.",
    "- Do not use a hardcoded service table, brand mapping, icon database, region list, operating system list, or local fallback knowledge.",
    "- For the subscription names below, notes.value must be a concise service/site description unless the service purpose is truly unknowable.",
    ...missingNoteNames.map((name) => `  - ${name}`),
    "- Use only the original input/images, high-confidence public knowledge, and dynamic fields already present in the previous object: service name, website/domain, category, and stable tags.",
    "",
    "Original recognition prompt:",
    "<<<renewlet-original-prompt",
    originalUserPrompt,
    ">>>",
    "",
    "Previous JSON object:",
    "<<<renewlet-previous-json",
    JSON.stringify(previousObject, null, 2),
    ">>>",
  ].join("\n");
}

function formatTagsForPrompt(tags: readonly string[]): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const tag of tags) {
    const value = tag.trim();
    const key = value.toLowerCase();
    if (!value || seen.has(key)) continue;
    seen.add(key);
    out.push(`- ${value}`);
    if (out.length >= 200) break;
  }
  return out.length === 0 ? ["- (none)"] : out;
}

function formatConfigOptionsForPrompt(options: readonly AIRecognitionPromptConfigOption[]): string[] {
  if (options.length === 0) return ["- (none)"];
  return options.slice(0, 200).map((option) => {
    const parts = [
      `value=${option.value}`,
      `label=${option.label}`,
      `zh-CN=${option.zhCN}`,
      `en-US=${option.enUS}`,
    ];
    return `- ${parts.join(" | ")}`;
  });
}

File: packages/server/cmd/renewlet/ai_recognition_prompt.go
package main

// ai_recognition_prompt.go 从 shared prompt JSON 生成 Go 运行面的提示词。
//
// shared/data/ai-recognition-prompt.json 是 Docker 与 Cloudflare 的事实源；Go 只读取 embedded 副本，
// 不能在运行面里手写另一套输出规则或字段解释。
import (
	"bytes"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	appstatic "github.com/zhiyingzzhou/renewlet/packages/server/internal/static"
)

type aiRecognitionPromptSpec struct {
	Version        string                       `json:"version"`
	SchemaName     string                       `json:"schemaName"`
	System         []string                     `json:"system"`
	OutputContract []string                     `json:"outputContract"`
	FieldRules     []string                     `json:"fieldRules"`
	Examples       []aiRecognitionPromptExample `json:"examples"`
}

type aiRecognitionPromptExample struct {
	Input  string          `json:"input"`
	Output json.RawMessage `json:"output"`
}

type aiRecognitionConfigOption struct {
	Value string
	Label string
	ZhCN  string
	EnUS  string
}

type aiRecognitionConfigContext struct {
	Categories     []aiRecognitionConfigOption
	PaymentMethods []aiRecognitionConfigOption
	Tags           []string
}

var aiRecognitionPrompt = mustLoadAIRecognitionPrompt()

func mustLoadAIRecognitionPrompt() aiRecognitionPromptSpec {
	var spec aiRecognitionPromptSpec
	if err := json.Unmarshal(appstatic.AIRecognitionPrompt, &spec); err != nil {
		panic(fmt.Sprintf("load AI recognition prompt: %v", err))
	}
	if spec.Version == "" || spec.SchemaName == "" || len(spec.System) == 0 || len(spec.OutputContract) == 0 || len(spec.FieldRules) == 0 {
		panic("load AI recognition prompt: incomplete prompt spec")
	}
	return spec
}

func buildAIRecognitionSystemPrompt() string {
	return strings.Join(aiRecognitionPrompt.System, "\n")
}

func buildAIRecognitionUserPrompt(text string, timezone string, defaultCurrency string, imageCount int, locale appLocale, configContext aiRecognitionConfigContext) string {
	examples := make([]string, 0, len(aiRecognitionPrompt.Examples))
	for index, example := range aiRecognitionPrompt.Examples {
		examples = append(examples, strings.Join([]string{
			fmt.Sprintf("Example %d input:", index+1),
			example.Input,
			fmt.Sprintf("Example %d JSON output:", index+1),
			formatAIRecognitionPromptJSON(example.Output),
		}, "\n"))
	}
	userInput := strings.TrimSpace(text)
	if userInput == "" {
		userInput = "(no text input; use attached images)"
	}
	lines := []string{
		"Runtime context:",
		fmt.Sprintf("- Current date in user's timezone: %s", todayDateOnly(time.Now().UTC(), timezone)),
		fmt.Sprintf("- User timezone: %s", timezone),
		fmt.Sprintf("- User locale: %s", locale),
		fmt.Sprintf("- Default currency hint: %s", defaultCurrency),
		fmt.Sprintf("- Attached image count: %d", imageCount),
		"",
		"User context:",
		"Existing user tags:",
	}
	lines = append(lines, formatAIRecognitionTags(configContext.Tags)...)
	lines = append(lines,
		"",
		"Available Renewlet configuration options:",
		"Categories:",
	)
	lines = append(lines, formatAIRecognitionConfigOptions(configContext.Categories)...)
	lines = append(lines, "Payment methods:")
	lines = append(lines, formatAIRecognitionConfigOptions(configContext.PaymentMethods)...)
	lines = append(lines,
		"",
		"Task:",
		"- Extract subscriptions from the delimited user input below and any attached images.",
		"- Treat the dynamic user context as preferences and available options, not as subscription evidence.",
		"",
		"Output contract:",
	)
	lines = append(lines, prefixAIPromptRules(aiRecognitionPrompt.OutputContract)...)
	lines = append(lines, "", "Field rules:")
	lines = append(lines, prefixAIPromptRules(aiRecognitionPrompt.FieldRules)...)
	lines = append(lines, "", "Examples:", strings.Join(examples, "\n\n"), "", "User input:", "<<<renewlet-user-input", userInput, ">>>")
	return strings.Join(lines, "\n")
}

func buildAIRecognitionRepairUserPrompt(originalUserPrompt string, previousObject interface{}, missingNoteNames []string) string {
	// repair 只修补缺失备注，不允许引入本地品牌表或图标库；否则 Docker/Worker 会在同一输入上产生不同草稿。
	lines := []string{
		"Repair task:",
		"- The previous JSON object is structurally valid but some describable subscriptions have missing or unusable notes.",
		"- Regenerate the entire JSON object with the same output contract, field rules, and all subscriptions.",
		"- Do not use a hardcoded service table, brand mapping, icon database, region list, operating system list, or local fallback knowledge.",
		"- For the subscription names below, notes.value must be a concise service/site description unless the service purpose is truly unknowable.",
	}
	for _, name := range missingNoteNames {
		lines = append(lines, "  - "+name)
	}
	lines = append(lines,
		"- Use only the original input/images, high-confidence public knowledge, and dynamic fields already present in the previous object: service name, website/domain, category, and stable tags.",
		"",
		"Original recognition prompt:",
		"<<<renewlet-original-prompt",
		originalUserPrompt,
		">>>",
		"",
		"Previous JSON object:",
		"<<<renewlet-previous-json",
		aiRecognitionJSONText(previousObject),
		">>>",
	)
	return strings.Join(lines, "\n")
}

func formatAIRecognitionTags(tags []string) []string {
	out := []string{}
	seen := map[string]bool{}
	for _, tag := range tags {
		value := strings.TrimSpace(tag)
		key := strings.ToLower(value)
		if value == "" || seen[key] {
			continue
		}
		seen[key] = true
		out = append(out, "- "+value)
		if len(out) >= 200 {
			// 标签上下文只提供偏好提示；过多用户历史标签会稀释订阅证据并放大 prompt 成本。
			break
		}
	}
	if len(out) == 0 {
		return []string{"- (none)"}
	}
	return out
}

func formatAIRecognitionConfigOptions(options []aiRecognitionConfigOption) []string {
	if len(options) == 0 {
		return []string{"- (none)"}
	}
	limit := len(options)
	if limit > 200 {
		limit = 200
	}
	out := make([]string, 0, limit)
	for _, option := range options[:limit] {
		out = append(out, fmt.Sprintf("- value=%s | label=%s | zh-CN=%s | en-US=%s", option.Value, option.Label, option.ZhCN, option.EnUS))
	}
	return out
}

func prefixAIPromptRules(rules []string) []string {
	out := make([]string, 0, len(rules))
	for _, rule := range rules {
		out = append(out, "- "+rule)
	}
	return out
}

func formatAIRecognitionPromptJSON(raw json.RawMessage) string {
	var buf bytes.Buffer
	if err := json.Indent(&buf, raw, "", "  "); err == nil {
		return buf.String()
	}
	return string(raw)
}
